### PR TITLE
Fix #46 Support JSX comments

### LINKED ISSUE #46: how to comment JSX code?
Is it possible to comment JSX code using emac's `(comment-line)`?

### PR DESCRIPTION
**Pull request by wyuenho**
This PR implements a number of comment commands suitable for `rjsx-mode`. 

`rjsx-comment-dwim` generally does what you expect - tries really hard to keep your JSX code valid by deciding what kind of `comment-start` and `comment-end` delimiters to use. If point is on a top-level JSX element, it'll just comment the whole node with `/* */`. If point is on a child element or text node, the child is commented with `{/* */}`. If point is on or inside a JSX expression, the body of the JSX expression is commented with `/* */`. If point  is on or before a JSX attribute, the entire attribute is commented with `/* */`.

This PR also provides an implementation of `comment-region-function`, `uncomment-region-function` and `comment-quote-nested-function` to support `rjsx-comment-dwim`.

There's a shit ton of tests included. Generally, `rjsx-comment-dwim` will just do what you expect. In the few edge cases that it doesn't, you can cycle the comments delimiters until you get what you want, or by moving point one character before or after its current position before `M-;`

I think this PR is so good, I've remapped `comment-dwim` to `rjsx-comment-dwim` when `rjsx-mode` is activated.

P.S. This is some of the ugliest and nastiest code I've ever written, and it's by no means easy as there are a LOT of cases to take care of and involves quite a number of regular expression hacks. I really hate `newcomment` and these stupid `comment-` global variables.

### CODE REVIEW

**Comment by felipeochoa** [comment]
Thanks a lot for this! On a quick check it looks great. I might make a few small changes, but hope to merge this this week

**Comment by wyuenho** [comment]
@felipeochoa How does this look? This PR comes with extensive tests, can it be merged yet?

**Comment by felipeochoa** [comment]
@wyuenho I'm sorry this has taken so long. I've been slammed with work and haven't had a chance to review any OSS stuff. I'll leave a couple of nitpicky comments right now that will help me do a proper review faster later on when I have a chance. (I usually just make such minor changes myself, but if you don't mind updating the PR for these, it will save some time)

File: rjsx-mode.el [hunk 1]
```diff
@@ -45,15 +47,6 @@ parsing supports the magic `rjsx-electric-lt' and
   :group 'rjsx-mode
   :type 'integer)
 
-;;;###autoload
```
Comment (felipeochoa) [inline rjsx-mode.el]: Can you move this part back up here? Want to keep the diff small and git blame working properly

File: rjsx-mode.el [hunk 2]
```diff
@@ -870,8 +863,6 @@ inserts `</>' and places the cursor inside the new tag."
                  (backward-char 2))
         (insert "<"))))
 
-(define-key rjsx-mode-map "<" 'rjsx-electric-lt)
```
Comment (felipeochoa) [inline rjsx-mode.el]: Ditto with this diff

File: rjsx-mode.el [hunk 3]
```diff
@@ -907,15 +896,13 @@ slash and inserts a matching end-tag."
   (interactive "p")
   (if (or killflag (/= 1 n) (not (eq (get-char-property (point) 'rjsx-class) 'self-closing-slash)))
       (if (called-interactively-p 'any)
-	  (call-interactively 'delete-forward-char)
```
Comment (felipeochoa) [inline rjsx-mode.el]: It's really gross that a tab got in here, but please revert this. I'd like to minimize the diff

File: rjsx-mode.el [hunk 4]
```diff
@@ -943,7 +930,269 @@ NEW-NAME is the name to give the tag."
           (save-excursion (goto-char beg) (insert new-name)))
         (js2-reparse))))))
 
-(define-key rjsx-mode-map (kbd "C-c C-r") 'rjsx-rename-tag-at-point)
+;; Comment handling
+(defun rjsx-ancester (predicate node)
```
Comment (felipeochoa) [inline rjsx-mode.el]: Two nits: 
1. `s/ancester/ancestor/` (there are a lot of these)
2. Can you rename this as `rjsx-find-ancestor`?

File: rjsx-mode.el [hunk 5]
```diff
@@ -943,7 +930,269 @@ NEW-NAME is the name to give the tag."
...
+  :lighter ":RJSX"
+  :group 'rjsx-mode
+  (setq-local comment-use-syntax nil)
+  (setq-local comment-start-skip "[[:space:]]*\\(//+\\|{?/\\*+\\)")
+  ;; \n is included to get arround `comment-normalize-vars' and `comment-only-p'
+  (setq-local comment-end-skip "\\(\\*+/}?[[:space:]]*\\)\n?\\|\n")
+  (setq-local comment-region-function 'rjsx-comment-region-function)
+  (setq-local uncomment-region-function 'rjsx-uncomment-region-function)
+  (setq-local comment-quote-nested-function 'rjsx-comment-quote-nested-function)
+  (define-key rjsx-mode-map [remap comment-dwim] 'rjsx-comment-dwim)
```
Comment (felipeochoa) [inline rjsx-mode.el]: Please define keys outside the major mode init function. This gets called every time `rjsx` is activated, but we only need to define it once.

Also, is this how other modes handle `comment-dwim`? I would have thought that it provided some hook of its own

File: rjsx-tests.el [hunk 6]
```diff
@@ -876,14 +876,14 @@ Currently only forms with syntax errors are supported.
       (dolist (contents cases)
         (insert (car contents))
         (save-excursion (insert (cdr contents)))
-	(js2-mode--and-parse)
```
Comment (felipeochoa) [inline rjsx-tests.el]: This tab thing is a real mystery, but please keep the diff focused

File: rjsx-tests.el [hunk 7]
```diff
@@ -1017,4 +1017,1070 @@ Currently only forms with syntax errors are supported.
...
+  (search-forward "<s"))
+
+(rjsx-deftest-comment
+  uncomment-dwim-nested-element
+  "(<div>{/* <span>hello</span> */}</div>)"
+  "(<div><span>hello</span></div>)"
+  'rjsx-comment-dwim
+  (search-forward "{"))
+
+(rjsx-deftest-comment
```
Comment (felipeochoa) [inline rjsx-tests.el]: Can you split this (and a couple of other tests below) into different tests. One for `/* className={1} */`, one for `/* namespace:foo=\"\" */`, etc.

**Comment by felipeochoa** [comment]
@wyuenho Also, thanks again for submitting this PR!! This is awesome functionality, and I'm really sorry I haven't had time to properly review

File: rjsx-mode.el (hunk 1 above)
Comment (wyuenho) [reply rjsx-mode.el]: No. The definition of rjsx-mode now has a body, I'd like the references in the body to have been already defined when invoked.

Comment (felipeochoa) [reply rjsx-mode.el]: Do you get compiler warnings if you put it up here?

File: rjsx-mode.el (hunk 5 above)
Comment (wyuenho) [reply rjsx-mode.el]: Org mode does this. AFAIK, no other mode remap comment-dwim, rjsx-mode is special cased because it operates on the AST as opposed to just a line or region, which in that case you can just redefine the hooks as soon above.

File: rjsx-mode.el (hunk 1 above)
Comment (wyuenho) [reply rjsx-mode.el]: No, it's a readability thing. I'll put this back. There's another easy peacy PR for a compiler warning at #73

**Comment by wyuenho** [comment]
@felipeochoa Should be good to go

File: rjsx-tests.el (hunk 7 above)
Comment (wyuenho) [reply rjsx-tests.el]: Done

**Comment by wyuenho** [comment]
I just realized I missed ancester again, it's all ancestor now.

**Comment by felipeochoa** [comment]
Ugh I was trying to make some small changes in [jsx-comments](https://github.com/felipeochoa/rjsx-mode/tree/jsx-comments) but couldn't figure out how to update the PR directly. 

I think I'm done reviewing and just have a couple substantive questions:

1. What is `rjsx-maybe-unwrap-expr` doing?
2. (Possibly related to above) Why do we need to have a custom `quote-nested` function?
3. Does JSX text commenting round-trip correctly? It seems like it would come back as a broken expression. Perhaps we should string-wrap when commenting it out?

My only concern with this code (related to point 2 I think) is treating `{/*` and `*/}` as a comment delimiters in some of the `comment-xxx` functions/variables. We need to be careful that there aren't issues with regular JS blocks with comments. E.g.:

```js
function noop() {/* pass */}
```

There are a few smaller notes I've left about code style. Thanks again for this great PR and sorry it's taken so long to review properly!

File: rjsx-mode.el [hunk 8]
```diff
@@ -945,6 +954,245 @@ NEW-NAME is the name to give the tag."
 
 (define-key rjsx-mode-map (kbd "C-c C-r") 'rjsx-rename-tag-at-point)
 
+;; Comment handling
+(defun rjsx-ancestor (predicate node)
```
Comment (felipeochoa) [inline rjsx-mode.el]: Can you flip the arguments here and fix the docstring so checkdoc doesn't complain? (I made this change in my jsx-comments branch already)

**Comment by wyuenho** [comment]
> 1. What is rjsx-maybe-unwrap-expr doing?

Uncommenting that call in `rjsx-uncomment-region-function` fails 2 tests. 

in `uncomment-dwim-text-vertical, uncommenting the second line when point is at BOL, without unwrapping will give you,

```jsx
  <div>
  |  {/* <span>hello</span> */}
  </div>
```
will give you

```
  <div>
  |  {<span>hello</span>}
  </div>
```
because the node `{/* ... */}` is treated as an empty JSX expression, so it was first uncommented as a JSX expression. While leaving `{<span>...</span>}` is perfectly fine, it's not really DWIM, so I just check if the presumably single node inside the JSX expression can be lifted up without changing it's meaning.

The test `rjsx-uncomment-dwim-nested-element` tests unwrapping for a similar reason, it's just this time the point is inside the JSX expression.

> 2. (Possibly related to above) Why do we need to have a custom quote-nested function?

No, a custom `comment-quote-nested-function` is needed because I switch `comment-start-skip` and `comment-end-skip` and `comment-start` and `comment-end` depending on context. The default `comment-quote-nested-function` naively assumes they never change. You can comment that line out in `rjsx-mode` 's body and take a look at the 4 failing tests to find out why.

> 3. Does JSX text commenting round-trip correctly? It seems like it would come back as a broken expression. Perhaps we should string-wrap when commenting it out?

Well, for JSX text, yes, the round trip works, but for uncommenting, this implementation doesn't distinguish between

```jsx
<div>
  {/* hello world */}
</div>
```

and 

```jsx
<div>
  {/* [1,2,3].map(num => num) */}
</div>
```
in both cases, they are unwrapped when uncommenting. So in the second case, the JSX expr will be broken, a js2-error-node results and further commenting will switch back to /* */. Dealing with this case is possible, basically I'll need to take the body of the only comment node in a JSX expression, throw it into a JSX node in a hidden buffer, and turn on `rjsx-mode` in it and see if it gets parsed as a JS Arithmetic Expression. I'll fix this after this PR is merged. This PR is blocking #75 .

> function noop() {/* pass */}

This is perfectly fine. Special handling for comments only happens inside a `rjsx-node` boundary. If that snippet in a `rjsx-node` isn't wrapped inside a JSX expr, it'll be treated as JSX text, and escaping and unescaping /* */ when commenting and uncommenting will work as expected. This is what the custom `comment-quote-nested-function` is used for. The problem is overzealous unwrapping as said above.

**Comment by wyuenho** [comment]
Actually, you are right, `function noop() {/* pass */}` outside of a JSX context is broken, I'll fix that. Good catch.

**Comment by bryzettler** [comment]
Any update on getting this and https://github.com/felipeochoa/rjsx-mode/pull/75 merged in?
Would love to have those indentation fixes.

**Comment by wyuenho** [comment]
I think this PR can be merged now. Any other bugs found can be fixed fairly easily in a seperately PR, I just haven't got the time to deal with the issue raised above, but it's should be hard, and anyone wants to fix it should be able to preempt me.

My preference is to merge this PR now so it doesn't block #75.

**Comment by felipeochoa** [comment]
Sorry if I'm being dumb here, but this is my first time using the github review system. Did you already address the comments I left there? I think they're all easy to fix. Happy to merge once those are addressed and file a new issue for this bug.

**Comment by wyuenho** [comment]
As I said, the issue with `function noop() {/* pass */}` will be fixed when I have time, possibly next week. This PR is not perfect, but it's pretty close to it now. There isn't any other issue that needs fixing.

**Comment by wyuenho** [comment]
@felipeochoa You can do all these yourself right? You can directly push to my branch and the result will show up here, or you can merge this PR first and then bikeshed later. I really don't have time to spend on this anymore other than `function noop() {/* pass */}`. This whole PR has been one clusterfuck of regex magic and tree traversals, it's not going to be pretty however you put lipstick on it.

File: rjsx-mode.el [hunk 9]
```diff
@@ -945,6 +954,245 @@ NEW-NAME is the name to give the tag."
...
+   (lambda ()
+     (let* ((node (js2-node-at-point beg))
+            (in-jsx (or (rjsx-node-p node)
+                        (rjsx-ancestor 'rjsx-node-p node)))
+            (use-jsx-comment (and (rjsx-node-p (js2-node-parent node))
+                                  (or (rjsx-text-p node)
+                                      (and (rjsx-node-p node)
+                                           (= (js2-node-abs-pos node) beg))))))
+       (cond (use-jsx-comment
+              (let ((comment-start "{/*")
```
[the remark this answers is not in the document]
Comment (wyuenho) [reply rjsx-mode.el]: There isn't anything that can be generalized here besides the `comment-start` and `comment-end`. You can define a `defsubst` if you want to have a function that you can read out loud and also remove the function call overhead.

File: rjsx-mode.el [hunk 10]
```diff
@@ -945,6 +954,245 @@ NEW-NAME is the name to give the tag."
...
+  (let ((ancestor (js2-node-parent node)))
+    (while (and ancestor
+                (not (funcall predicate ancestor)))
+      (setq ancestor (js2-node-parent ancestor)))
+    ancestor))
+
+(defun rjsx-comment-region-function (beg end &optional arg)
+  (js2-mode-wait-for-parse
+   (lambda ()
+     (let* ((node (js2-node-at-point beg))
```
[the remark this answers is not in the document]
Comment (wyuenho) [reply rjsx-mode.el]: I don't see how inlining a giant blob of parens into the conditions is going to make this code more readable. You can have (a) or (b), but not both.

File: rjsx-mode.el [hunk 11]
```diff
@@ -945,6 +954,245 @@ NEW-NAME is the name to give the tag."
...
+         (setq ts (match-end 1))
+         (goto-char cs)
+         (delete-region cs ts)
+
+         ;; delete comment-padding start
+         (when (and comment-padding (looking-at (regexp-quote comment-padding)))
+           (delete-region (point) (+ (point) (length comment-padding))))
+
+         ;; find comment end
+         (when (re-search-forward (if (string-match "//+" matched-start) "\n" "\\*/}?") end t 1)
```
[the remark this answers is not in the document]
Comment (wyuenho) [reply rjsx-mode.el]: Pretty sure `forward-comment` won't work as it matches using the syntax table. We do need to match strings here.

If you are trying to fix `function noop() {/* pass */}`, the solution should probably involve a private global variable to communicate between `rjsx-comment-dwim` and `rjsx-uncomment-region-function`, and switch between comment delimiters depending on whether you are in JSX or not, but obviously this is harder than it seems.

File: rjsx-mode.el [hunk 12]
```diff
@@ -945,6 +954,245 @@ NEW-NAME is the name to give the tag."
...
+           (save-restriction
+             (narrow-to-region cs (point))
+             (comment-quote-nested "{/*" "*/}" t)))))
+
+     (rjsx-maybe-unwrap-expr beg end)
+
+     (set-marker end nil))))
+
+(defun rjsx-comment-quote-nested-function (_ __ unp)
+  (let ((re (concat "\\*\\(\\\\" (if unp "+" "*") "\\)/}?"
```
[the remark this answers is not in the document]
Comment (wyuenho) [reply rjsx-mode.el]: This regex is almost entirely the same as `comment-quote-nested-function`'s. I'd rather not change it too much in case we need to reference back to it. An extra layer of syntax translation doesn't really help comprehension here.

File: rjsx-mode.el [hunk 13]
```diff
@@ -945,6 +954,245 @@ NEW-NAME is the name to give the tag."
...
+      (goto-char beg)
+      (skip-chars-forward "[:space:]\n" end)
+      (let* ((node (js2-node-at-point (point)))
+             (parent (js2-node-parent node)))
+        (when (and parent
+                   (rjsx-wrapped-expr-p parent)
+                   (rjsx-node-p (js2-node-parent parent))
+                   (or (rjsx-text-p node) (rjsx-node-p node)))
+          (let* ((expr-start (js2-node-abs-pos parent))
+                 (expr-end (js2-node-abs-end parent))
```
[the remark this answers is not in the document]
Comment (wyuenho) [reply rjsx-mode.el]: Good to know!

**Comment by felipeochoa** [comment]
Shoot so I apparently never hit submit on the changes I was requesting...

**Comment by felipeochoa** [comment]
@wyuenho I appreciate you putting in the time to create this (and the other!) PR, and I'm very sorry I haven't had much time to engage. That said, I do want to understand how this whole thing comes together since I'm going to be maintaining this code for the foreseeable future. Hence some amount of lipstick is necessary :)

The fact that there is so much regex magic when we have an AST is a little concerning. What is the point of having an AST if we're going to fall back to regexes anyway?

Re: bike-shedding. I thought I tried pushing changes earlier (weeks ago) but was rejected -- did you enable the "allow maintainers to edit" or whatever it's called flag? (Totally possible it's my fault)

**Comment by wyuenho** [comment]
But you do understand how all this come together, even better than I do :)

The regex in the uncommenting code is necessary is because trying to parse a comment code's content by throwing every one of them in a temp buffer and turn on rjsx-mode and thus the js2 parser is insane. There's that and have to interface with those 20-year-old "newcomment" hooks, and that JSX comment isn't actually a part of the syntax defined anywhere, and there are no AST classes in rjsx to help me.

But it all works! Miraculously. 

>  thought I tried pushing changes earlier (weeks ago) but was rejected 

screen shot 2018-07-20 at 01 52 19

You were probably trying to push after I've rebased the branch. I promise I won't do it anymore :)

P.S. It surprises me that I can still find someone on Github who doesn't quite know how to use it :) What do you use normally?

**Comment by anselstetter** [comment]
So I guess this, and #75 won't get merged?

**Comment by wyuenho** [comment]
@felipeochoa Merge now?

**Comment by felipeochoa** [comment]
I was actually thinking about this PR today... I was going to suggest I just merge them and then ping you if any issues come up with it.
